src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp

```diff
@@ -0,0 +1,91 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//==== ONNXToTosaLegalizeUtils.hpp - ONNX dialects to TOSA lowering Utils-===//
+//
+// Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+// Copyright (c) 2022 Advanced Micro Devices, Inc.
+//
+// =============================================================================
+//
+// This file contains common utils shared by the functions performing the
+// lowering to the TOSA dialect. It is also used by TensorFlow and torch-mlir.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef ONNXMLIR_CONVERSION_ONNXTOTOSA_TOSALEGALIZEUTILS_H
+#define ONNXMLIR_CONVERSION_ONNXTOTOSA_TOSALEGALIZEUTILS_H
+
+#include "mlir/Dialect/Quant/QuantTypes.h"        // from @llvm-project
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"         // from @llvm-project
+#include "mlir/Dialect/Tosa/Utils/ShapeUtils.h"   // from @llvm-project
+#include "mlir/IR/BuiltinAttributes.h"            // from @llvm-project
+#include "mlir/IR/BuiltinTypes.h"                 // from @llvm-project
+#include "mlir/IR/PatternMatch.h"                 // from @llvm-project
+#include "mlir/Interfaces/InferTypeOpInterface.h" // from @llvm-project
+#include "mlir/Support/LLVM.h"                    // from @llvm-project
+
+namespace onnx_mlir {
+namespace tosa {
+
+// Creates a TOSA operation and performs shape inference on the individual
+// op. This allows shape inference during the framework to TOSA lowering.
+template <typename TosaOp, typename... Args>
+TosaOp CreateOpAndInfer(mlir::PatternRewriter &rewriter, mlir::Location loc,
+    mlir::Type result_ty, Args &&... args) {
+  auto op = rewriter.create<TosaOp>(loc, result_ty, args...);
+
+  mlir::InferShapedTypeOpInterface shapeInterface =
+      llvm::dyn_cast<mlir::InferShapedTypeOpInterface>(op.getOperation());
+  if (!shapeInterface)
+    return op;
+
+  llvm::SmallVector<mlir::ShapedTypeComponents> returnedShapes;
+  if (shapeInterface
+          .inferReturnTypeComponents(op.getContext(), op.getLoc(),
+              op->getOperands(), op->getAttrDictionary(), op->getRegions(),
+              returnedShapes)
+          .failed())
```

Comment:
Does `inferReturnTypeComponents` generate runtime code for unknown dimensions or just return -1 if unknown?

@AlexandreEichenberger fyi, `InferShapedTypeOpInterface` may be interesting to you. There is a duplicated effort here to infer shape and update shape.

Reply:
It depends on the specific implementation of the interface method for each operation. Most of the time it returns a dynamic shape if it cannot figure out the right one.